File: haskell/pairing.hs
import qualified Data.ByteString.Char8 as B
import Control.Monad
import Data.Maybe
 
data Heap a = Empty | Heap a [Heap a]
 
link :: Ord a => Heap a -> Heap a -> Heap a
link Empty heap = heap
link heap Empty = heap
link lhs@(Heap lhsval lhssons) rhs@(Heap rhsval rhssons)
  | lhsval < rhsval = Heap lhsval (rhs : lhssons)
  | otherwise       = Heap rhsval (lhs : rhssons)
 
merge :: Ord a => [Heap a] -> Heap a
merge [] = Empty
merge [heap] = heap
merge (h1 : h2 : hs) = link (link h1 h2) (merge hs)
 
makeheap :: a -> Heap a
makeheap dat = Heap dat []
 
top :: Heap a -> a
top Empty = error "Empty heap"
top (Heap val _) = val
 
pop :: Ord a => Heap a -> Heap a
pop Empty = error "Empty heap"
pop (Heap _ sons) = merge sons
 
insert :: Ord a => a -> Heap a -> Heap a
insert val heap = link heap (makeheap val)
 
process :: Heap Int -> B.ByteString -> IO (Heap Int)
process heap request = 
  case action of
    0 -> return (insert value heap)
    1 -> do
          print $ top heap
          return heap
    2 -> return $ pop heap
  where
    (action, rest) = fromJust $ B.readInt request
    (value, _)  = fromJust $ B.readInt $ B.drop 1 rest
 
main :: IO ()
main = do
  contents <- B.getContents
  foldM_ process (Empty::(Heap Int)) (B.lines contents)
 

File: ocaml/pairing.ml
open Scanf

type 'a heap = Empty | Heap of 'a * 'a heap list

let link h1 h2 = match h1, h2 with
  | Empty, heap -> heap
  | heap, Empty -> heap
  | Heap (k1, c1), Heap (k2, c2) when k1 < k2 ->
      Heap (k1, h2 :: c1)
  | Heap (k1, c1), Heap (k2, c2) ->
      Heap (k2, h1 :: c2);;

let rec merge heap = match heap with
  | [] -> Empty
  | [heap] -> heap
  | h1 :: h2 :: hs -> link (link h1 h2) (merge hs);;

let makeheap key = Heap (key, []);;

let top heap = match heap with
  | Empty -> failwith "Empty heap"
  | Heap (key, _) -> key;;

let pop heap = match heap with
  | Empty -> failwith "Empty heap"
  | Heap (_, chld) -> merge chld;;

let insert heap key = link heap (makeheap key);;

let rec process rheap =
  scanf "%d" (fun mode ->
    match mode with
      | 0 -> scanf " %d\n" (fun key ->
              rheap := insert !rheap key)
      | 1 -> scanf "\n" ignore; print_endline (string_of_int (top !rheap))
      | 2 -> scanf "\n" ignore; rheap := pop !rheap);
  process rheap;;

try
  process (ref Empty)
with End_of_file -> ();;
